### PR TITLE
Featureimage/pqswmbt 4894 poc mqtt

### PR DESCRIPTION
**Pull request by MimiXav**

### CODE REVIEW

**Comment by etnKlendathu** [comment]
I have a few questions, comments first :)
1) Why is it common message bus... do we have uncommon one? :)
2) Why do you hate move semantic? :)
3) This is a pity that we have no common style (not only in code, but if file naming too) f.e. case sensitive names, hpp as extensions etc... this code looks very alien if compare it with other fty-* projects
4) This change is too big to review in one piece... how to process it?

File: amqp/src/AmqpClient.cpp [hunk 1]
```diff
@@ -19,160 +19,287 @@
...
 namespace fty::messagebus::amqp
 {
-  static auto constexpr AMQP_CORREL_ID = "JMSCorrelationID";
+  using namespace fty::messagebus;
   using MessageListener = fty::messagebus::MessageListener;
-  using fty::messagebus::Message;
 
-  AmqpClient::AmqpClient(const std::string& url, const std::string& address, const std::string& filter, MessageListener messageListener)
+  static auto constexpr TIMEOUT = std::chrono::seconds(5);
+  static auto constexpr AMQP_CORREL_ID = "JMSCorrelationID";
```
Comment (jeffdegott) [inline amqp/src/AmqpClient.cpp:54]: AMQP_ prefix is useless here

File: amqp/src/AmqpClient.cpp [hunk 2]
```diff
@@ -19,160 +19,287 @@
...
 
   void AmqpClient::on_container_start(proton::container& container)
   {
-    logDebug("AmqpClient on_container_start");
     try
     {
-      container.connect(m_url);
+      container.connect(m_url, connectOpts().reconnect(reconnectOpts()));
     }
     catch (std::exception& e)
```
Comment (jeffdegott) [inline amqp/src/AmqpClient.cpp:73]: const std::exception &e

File: amqp/src/AmqpClient.cpp [hunk 3]
```diff
@@ -19,160 +19,287 @@
...
+  DeliveryState AmqpClient::receive(const Address& address, const std::string& filter, MessageListener messageListener)
+  {
+    auto deliveryState = DeliveryState::DELIVERY_STATE_REJECTED;
+    if (connected() == ComState::COM_STATE_OK)
+    {
+      logDebug("Set receiver to wait message(s) from {} ...", address);
+      m_promiseReceiver = std::promise<void>();
+
+      auto futureReceiver = m_promiseReceiver.get_future();
+      (!filter.empty())? setSubscriptions(filter, messageListener) : setSubscriptions(address, messageListener);
```
Comment (jeffdegott) [inline amqp/src/AmqpClient.cpp:188]: old style C... i prefer with an ' if '

File: amqp/src/AmqpClient.cpp [hunk 4]
```diff
@@ -19,160 +19,287 @@
...
+      if (futureReceiver.wait_for(TIMEOUT) != std::future_status::timeout)
+      {
+        deliveryState = DeliveryState::DELIVERY_STATE_ACCEPTED;
+      }
+    }
+
+    return deliveryState;
   }
 
   bool AmqpClient::tryConsumeMessageFor(std::shared_ptr<proton::message> resp, int timeout)
```
Comment (jeffdegott) [inline amqp/src/AmqpClient.cpp:203]: fir readability, here timeout shall be typed as chrono::seconds....
or at least named timeout_sec....

File: amqp/src/MsgBusAmqp.cpp [hunk 5]
```diff
@@ -45,84 +36,89 @@ namespace fty::messagebus::amqp
...
       return fty::unexpected(to_string(ComState::COM_STATE_CONNECT_FAILED));
     }
-
     return {};
   }
 
-  fty::Expected<void> MsgBusAmqp::receive(const std::string& address, MessageListener messageListener, const std::string& filter)
+  bool MsgBusAmqp::isServiceAvailable()
+  {
+    return (m_amqpClient && (m_amqpClient->connected() == ComState::COM_STATE_OK)) ? true : false;
```
Comment (jeffdegott) [inline amqp/src/MsgBusAmqp.cpp:75]: " ? true : false" is useless

File: amqp/tests/MessageBusAmqpTest.cpp [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
+      auto msgBusReplyer = amqp::MessageBusAmqp("AsyncReplyerTestCase", AMQP_SERVER_URI);
+      REQUIRE(msgBusReplyer.connect());
+
+      // Build asynchronous request and set all receiver
+      Message request = Message::buildRequest("AsyncRequestTestCase", asyncTestQueue + "request", "TEST", asyncTestQueue + "reply", QUERY);
+      REQUIRE(msgBusReplyer.receive(request.to(), std::bind(&MsgReceived::replyerAddOK, std::ref(msgReceived), std::placeholders::_1)));
+      REQUIRE(msgBusRequester.receive(request.replyTo(), std::bind(&MsgReceived::messageListener, std::ref(msgReceived), std::placeholders::_1), request.correlationId()));
+
+      for (int i = 0; i < 1; i++)
```
Comment (jeffdegott) [inline amqp/tests/MessageBusAmqpTest.cpp:214]: loop is useless?

File: amqp/tests/MessageBusAmqpTest.cpp [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+      auto request = Message::buildRequest("WrongRequestTestCase", "", "SyncTest", "", QUERY);
+      // Request reject
+      REQUIRE(msgBus.request(request, 1).error() == to_string(DeliveryState::DELIVERY_STATE_REJECTED));
+      request.from("queue://etn.q.request");
+      request.to("queue://etn.q.reply");
+      // Without reply request reject.
+      REQUIRE(msgBus.request(request, 1).error() == to_string(DeliveryState::DELIVERY_STATE_REJECTED));
+    }
+
+    SECTION("Wrong message")
```
Comment (jeffdegott) [inline amqp/tests/MessageBusAmqpTest.cpp:315]: same section name ?

File: common/public_include/fty/messagebus/Message.h [hunk 8]
```diff
@@ -27,18 +27,20 @@ namespace fty::messagebus
...
-  static constexpr auto SUBJECT = "SUBJECT";
-  static constexpr auto STATUS = "STATUS";
-  static constexpr auto TIMEOUT = "TIMEOUT";
+  static constexpr auto MESSAGE_ID     = "MESSAGE_ID";
+  static constexpr auto FROM           = "FROM";         //ClientId of the message
+  static constexpr auto TO             = "TO";             //Queue of destination
+  static constexpr auto REPLY_TO       = "REPLY_TO"; //Queue for the reply
+  static constexpr auto SUBJECT        = "SUBJECT";
+  static constexpr auto STATUS         = "STATUS";
+  static constexpr auto TIMEOUT        = "TIMEOUT";
```
Comment (jeffdegott) [inline common/public_include/fty/messagebus/Message.h:43]: timeout sec? msec?   (add a comment)

File: etnWrapper/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,26 @@
+project(etn-common-messagebus2
+        VERSION ${PROJECT_VERSION}
+        DESCRIPTION "Eaton messagebus2"
```
Comment (jeffdegott) [inline etnWrapper/CMakeLists.txt:3]: Eaton VS. 42ity?

File: etnWrapper/public_include/etn/messagebus/EtnMessageBus.h [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    {
+      return m_mqttEndpoint;
+    }
+
+  private:
+    fty::messagebus::Endpoint m_amqpEndpoint;
+    fty::messagebus::Endpoint m_mqttEndpoint;
+  };
+
+  class EtnMessageBus final : public fty::messagebus::MessageBus
```
Comment (jeffdegott) [inline etnWrapper/public_include/etn/messagebus/EtnMessageBus.h:85]: classname 'Etn...' is surprising

**Review by FrancoisRegisDegott-eaton — DISMISSED** [review]

File: amqp/public_include/fty/messagebus/amqp/MessageBusAmqp.h [hunk 11]
```diff
@@ -26,29 +26,31 @@
...
   {
   public:
     MessageBusAmqp(const ClientName& clientName = utils::getClientId("MessageBusAmqp"),
-                   const Endpoint& endpoint = DEFAULT_AMQP_END_POINT);
+                   const Endpoint& endpoint = DEFAULT_ENDPOINT);
 
     ~MessageBusAmqp() = default;
 
-    MessageBusAmqp(MessageBusAmqp&& other) = default;
+    MessageBusAmqp(MessageBusAmqp&& other) = delete;
```
Comment (etnKlendathu) [inline amqp/public_include/fty/messagebus/amqp/MessageBusAmqp.h:45]: Why???? :)

File: amqp/src/AmqpClient.cpp [hunk 12]
```diff
@@ -19,160 +19,286 @@
...
-    logDebug("Checking answer for {} second(s)...", timeout);
+    auto deliveryState = DeliveryState::DELIVERY_STATE_REJECTED;
+    if (connected() == ComState::COM_STATE_OK)
+    {
+      m_promiseSender = std::promise<void>();
+      logDebug("Sending message to {} ...", msg.to());
+      m_message.clear();
+      m_message = msg;
+
+      m_connection.work_queue().add([=]() {
```
Comment (etnKlendathu) [inline amqp/src/AmqpClient.cpp:165]: Why do you capture by copy here if you store message as instance property?

File: amqp/src/AmqpClient.cpp [hunk 13]
```diff
@@ -19,160 +19,286 @@
...
-    // The value is a specific AMQP "described type": binary string with symbolic descriptor
-    proton::codec::encoder enc(filterValue);
-    enc << proton::codec::start::described()
-        << proton::symbol("apache.org:selector-filter:string")
-        << selector
-        << proton::codec::finish();
-    // In our case, the map has this one element
-    map.put(filterKey, filterValue);
-    opts.filters(map);
+    if (auto it{m_subscriptions.find(address)}; it == m_subscriptions.end() && messageListener)
```
Comment (etnKlendathu) [inline amqp/src/AmqpClient.cpp:271]: IMHO more natural will be if (auto it = m_subscriptions.find(); ut !=....)

File: amqp/src/AmqpClient.h [hunk 14]
```diff
@@ -30,48 +31,60 @@
...
-    bool tryConsumeMessageFor(std::shared_ptr<proton::message> resp, int timeout);
-    void send(const proton::message& msg);
+    void on_error(const proton::error_condition& error) override;
+    void on_transport_error(proton::transport &t) override;
+
+    fty::messagebus::ComState connected();
+    fty::messagebus::DeliveryState receive(const Address& address, const std::string& filter = {}, MessageListener messageListener = {});
+    fty::messagebus::DeliveryState unreceive();
+    fty::messagebus::DeliveryState send(const proton::message& msg);
+    bool tryConsumeMessageFor(std::shared_ptr<proton::message> resp, int timeoutInSeconds);
```
Comment (etnKlendathu) [inline amqp/src/AmqpClient.h:64]: Why do you use different code styles in one class? hungarian and camel cases?

File: amqp/src/MsgBusAmqp.h [hunk 15]
```diff
@@ -58,11 +63,14 @@ namespace fty::messagebus::amqp
       return m_clientName;
     }
 
+    bool isServiceAvailable();
+
   private:
     std::string m_clientName{};
```
Comment (etnKlendathu) [inline amqp/src/MsgBusAmqp.h:69]: You don't need {} here. Default ctor by default.

File: amqp/src/MsgBusAmqp.h [hunk 16]
```diff
@@ -58,11 +63,14 @@ namespace fty::messagebus::amqp
...
-    std::string m_endpoint{};
```
Comment (etnKlendathu) [inline amqp/src/MsgBusAmqp.h:63]: You don't need {} here. Default ctor by default.

File: amqp/tests/MessageBusAmqpTest.cpp [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+        std::cerr << response.error() << std::endl;
+      }
+
+      //send the response
+      auto msgBus = amqp::MessageBusAmqp("TestCase", AMQP_SERVER_URI);
+      auto connected = msgBus.connect();
+      auto msgSent = msgBus.send(response.value());
+      if (!msgSent)
+      {
+        std::cerr << msgSent.error() << std::endl;
```
Comment (etnKlendathu) [inline amqp/tests/MessageBusAmqpTest.cpp:118]: Maybe better to use FAIL(msgSent.error());

File: amqp/tests/MessageBusAmqpTest.cpp [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+      // Send message on queue
+      Message msg = Message::buildMessage("MqttMessageTestCase", sendTestQueue, "TEST", QUERY);
+
+      int nbMessageToSend = 3;
+      for (int i = 0; i < nbMessageToSend; i++)
+      {
+        REQUIRE(msgBusSender.send(msg));
+      }
+      std::this_thread::sleep_for(TIMEOUT);
+      CHECK(msgReceived.isRecieved(nbMessageToSend));
```
Comment (etnKlendathu) [inline amqp/tests/MessageBusAmqpTest.cpp:158]: I don't like this approach... maybe better to use conditinal_variable here... timeout could be to long and too short in some cases... :(

File: common/src/Message.cpp [hunk 19]
```diff
@@ -223,4 +232,12 @@ namespace fty::messagebus
     return data;
   }
 
+  Message& Message::operator=(const Message& other)
+  {
+    m_metadata = std::move(other.metaData());
+    m_data = std::move(other.userData());
+    return *this;
+  }
```
Comment (etnKlendathu) [inline common/src/Message.cpp:240]: use const Message& other (const reference) and after that move?

File: etnWrapper/src/EtnMessageBus.cpp [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+{
+  using namespace fty::messagebus;
+  using namespace fty::messagebus::amqp;
+  using namespace fty::messagebus::mqtt;
+
+  static const std::string INDENTITY("etnMessageBusAmqpMqtt");
+
+  fty::Expected<void> EtnMessageBus::send(const Message& msg) noexcept
+  {
+    fty::Expected<void> result;
```
Comment (etnKlendathu) [inline etnWrapper/src/EtnMessageBus.cpp:36]: Expected has no default ctor... this is a misusage...

File: etnWrapper/tests/EtnMessageBusTest.cpp [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+
+  //----------------------------------------------------------------------
+  // Test case
+  //----------------------------------------------------------------------
+
+  TEST_CASE("queue", "[etn][request][send]")
+  {
+    SECTION("Send")
+    {
+      MsgReceived msgReceived{};
```
Comment (etnKlendathu) [inline etnWrapper/tests/EtnMessageBusTest.cpp:134]: Again, no need explicitly call default ctor by {}

File: mqtt/CMakeLists.txt [hunk 22]
```diff
@@ -35,5 +35,7 @@ if(BUILD_TESTING)
   etn_test_target(${PROJECT_NAME}
     SOURCES
       tests/*.cpp
-  )
+    INCLUDE_DIRS
+      src/*.h
```
Comment (etnKlendathu) [inline mqtt/CMakeLists.txt:39]: *.h are not include dirs :)

File: mqtt/src/CallBack.cpp [hunk 23]
```diff
@@ -85,48 +85,35 @@ namespace fty::messagebus::mqtt
...
-
-  // Callback called for connection done.
-  void CallBack::onConnected(const std::string& cause)
-  {
-    logDebug("Connected");
-    if (!cause.empty())
-    {
-      logDebug("raison: {}", cause);
-    }
+    (cause.empty()) ? logError("Connection lost!") : logError("Connection lost: {}", cause);
```
Comment (etnKlendathu) [inline mqtt/src/CallBack.cpp:88]: Please normal if in this case.
or at least logError(cond? msg1: msg2);

File: mqtt/src/CallBack.cpp [hunk 24]
```diff
@@ -85,48 +85,35 @@ namespace fty::messagebus::mqtt
...
   }
 
-  // Callback called for connection updated.
-  bool CallBack::onConnectionUpdated(const ::mqtt::connect_data& /*connData*/)
+  auto CallBack::subscriptions() -> SubScriptionListener
```
Comment (etnKlendathu) [inline mqtt/src/CallBack.cpp:91]: You don't need return type detection. please use normal form rettype func()....

File: mqtt/src/CallBack.h [hunk 25]
```diff
@@ -47,12 +47,11 @@ namespace fty::messagebus::mqtt
...
     ~CallBack() = default;
     void connection_lost(const std::string& cause) override;
-    void onConnected(const std::string& cause);
-    bool onConnectionUpdated(const ::mqtt::connect_data& connData);
     void onMessageArrived(::mqtt::const_message_ptr msg, AsynClientPointer clientPointer = nullptr);
 
-    SubScriptionListener getSubscriptions();
-    void setSubscriptions(const std::string& topic, MessageListener messageListener);
+    SubScriptionListener subscriptions();
+    void subscriptions(const std::string& topic, MessageListener messageListener);
```
Comment (etnKlendathu) [inline mqtt/src/CallBack.h:53]: listener by copy?
Bellow you use universal semantic....

**Review by etnKlendathu — COMMENTED** [review]
I did just quick review of code... unfortunately this code need to be fixed :(

File: amqp/src/AmqpClient.cpp [hunk 26]
```diff
@@ -19,160 +19,286 @@
...
+  {
+    return proton::connection_options();
+  }
+
+} // namespace
+
 namespace fty::messagebus::amqp
 {
-  static auto constexpr AMQP_CORREL_ID = "JMSCorrelationID";
+  using namespace fty::messagebus;
```
[the remark this answers is not in the document]
Comment (MimiXav) [reply amqp/src/AmqpClient.cpp:50]: already removed

**Review by NavarroPierre — APPROVED** [review]

**Pull request merged**